caffe2/operators/onnxifi_op.cc

```diff
@@ -0,0 +1,192 @@
+      t->add_int32_data(*data++);
+    }
+  } else {
+    CAFFE_THROW(
+        "Don't know how to convert workspace tensor type ",
+        cpu_tensor.meta().name(),
+        " to ONNX TensorProto");
+  }
+}
+
+void BlobToTensorDescriptor(
+    const std::string& name,
+    Workspace* ws,
+    onnxTensorDescriptor* desc,
+    std::vector<std::vector<uint64_t>>* shapes) {
+  const Blob* blob = ws->GetBlob(name);
+  CAFFE_ENFORCE(blob, "Blob ", name, " doesn't exist");
+
+  // Memory type
+  // We only allow weights to be CPU tensor for now
+  CAFFE_ENFORCE(
+      blob->template IsType<TensorCPU>(),
+      "Initialization blob ",
+      name,
+      " needs to be TensorCPU");
+  desc->memoryType = ONNXIFI_MEMORY_TYPE_CPU;
+
+  // Data type
+  const auto& cpu_tensor = blob->template Get<TensorCPU>();
+  if (cpu_tensor.template IsType<float>()) {
+    desc->dataType = ONNXIFI_DATATYPE_FLOAT32;
+    desc->buffer = (onnxPointer)(cpu_tensor.data<float>());
+  } else if (cpu_tensor.template IsType<int64_t>()) {
+    desc->dataType = ONNXIFI_DATATYPE_INT64;
+    desc->buffer = (onnxPointer)(cpu_tensor.data<int64_t>());
+  } else if (cpu_tensor.template IsType<int32_t>()) {
+    desc->dataType = ONNXIFI_DATATYPE_INT32;
+    desc->buffer = (onnxPointer)(cpu_tensor.data<int32_t>());
+  }
+
+  // Set dims
+  const auto shape = GetTensorShapeOfBlob(blob);
+  desc->dimensions = shape.dims_size();
+  shapes->emplace_back();
+  auto& shape_tmp = shapes->back();
+  for (const auto d : shape.dims()) {
+    shape_tmp.push_back(d);
+  }
+  desc->shape = shape_tmp.data();
+}
+} // namespace
+
+template <>
+std::vector<onnxTensorDescriptor>
+OnnxifiOp<float, CPUContext>::BuildInitializationList(
+    Workspace* ws,
+    std::unordered_set<std::string>* initialization_list,
+    std::vector<std::string>* weight_names,
+    std::vector<std::vector<uint64_t>>* weight_shapes) {
+  const std::vector<string>& ws_blobs = ws->Blobs();
+  std::vector<onnxTensorDescriptor> descs;
+  for (const auto& s : ws_blobs) {
+    auto it = initialization_list->find(s);
+    if (it != initialization_list->end()) {
+      weight_names->emplace_back(s);
+      descs.emplace_back();
+      auto& tensor_desc = descs.back();
+      tensor_desc.name = weight_names->back().c_str();
+      BlobToTensorDescriptor(s, ws, &tensor_desc, weight_shapes);
+      initialization_list->erase(it);
+    }
+  }
+  CAFFE_ENFORCE(
+      initialization_list->empty(), "Unfulfilled initialization list");
+  return descs;
+}
+
+template <>
+bool OnnxifiOp<float, CPUContext>::RunOnDevice() {
+  for (unsigned i = 0U; i < InputSize(); ++i) {
+    const auto& input_tensor = Input(i);
+    const auto& tensor_dims = input_tensor.dims();
+    auto& tensor_descriptor = input_desc_.at(i);
+    tensor_descriptor.dataType = ONNXIFI_DATATYPE_FLOAT32;
+    tensor_descriptor.memoryType = ONNXIFI_MEMORY_TYPE_CPU;
+    tensor_descriptor.dimensions = tensor_dims.size();
+    input_shapes_.emplace_back();
```

Comment:
`input_shapes_.push_back(std::vector<uin64_t>(tensor_dims.cbegin(), tensor_dims.cend()));` would replace this and the four following lines